File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = Arrays.stream(nums).sum();
        if(sum % 2 == 1) return false;

        int val = sum / 2, n = nums.length;
        boolean[] dp = new boolean[val + 1]; dp[0] = true;

        for(int i = n - 1; i >= 0; i--){
            boolean[] curr = new boolean[val + 1]; curr[0] = true;
            for(int j = 1; j <= val; j++){
                boolean notPick = dp[j], pick = (j - nums[i] >= 0) ? dp[j - nums[i]] : false;
                curr[j] = pick || notPick;
            }
            dp = curr;
        }

        return dp[val];
    }
}

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
private:
    // Recursion
    bool fun(vector<int> &nums, int i, int target){
        if(i == 0){
            return nums[0] == target;
        }

        return  target == 0 || 
                fun(nums, i - 1, target) || 
                (target >= nums[i] ? fun(nums, i - 1, target - nums[i]) : false);
    }

public:
    bool canPartition(vector<int>& nums) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if(sum & 1){
            return false;
        }

        int n = nums.size();
        int target = sum / 2;
        // return fun(nums, n - 1, target);

        // Dp
        vector<bool> prev(target + 1, false);
        for(int val = 0; val <= target; val++){
            prev[val] = (val == nums[0]);
        }

        prev[0] = true;

        for(int i = 1; i < n; i++){
            vector<bool> curr(target + 1, false);
            curr[0] = true;
            for(int val = 1; val <= target; val++){
                curr[val] = prev[val] || (val >= nums[i] ? prev[val - nums[i]] : false);
            }
            prev = curr;
        }

        return prev[target];
    }
};